runtime/lua/vim/treesitter.lua

```diff
@@ -114,4 +114,106 @@ function M.get_string_parser(str, lang, opts)
   return LanguageTree.new(str, lang, opts)
 end
 
+--- Determines whether a node is the ancestor of another
+---
+---@param dest the possible ancestor
+---@param source the possible descendant node
+---
+---@returns True if dest is an ancestor of source
+function M.is_ancestor(dest, source)
+  if not (dest and source) then
+    return false
+  end
+
+  local current = source
+  while current ~= nil do
+    if current == dest then
+      return true
+    end
+
+    current = current:parent()
+  end
+
+  return false
+end
+
+--- Get the next node of the same parent
+---
+---@param node The start node
+---@param opts table A table of optional arguments:
+---                  - switch_parent: (boolean, default false) Allow to return the first node from
+---                  the next parent if the start node is the last node of its parent.
+---                  - return_next_parent: (boolean, default false) With switch_parent = true, allow
+---                  to return the next parent as the next node if it has no children.
+---
+---@returns The next node
+function M.get_next_node(node, opts)
+  local destination_node
+  local parent = node:parent()
+
+  if not parent then
+    return
+  end
+
+  opts = opts or {}
+
+  local found_pos = 0
+  for i = 0, parent:named_child_count() - 1, 1 do
+    if parent:named_child(i) == node then
+      found_pos = i
+      break
+    end
+  end
+  if parent:named_child_count() > found_pos + 1 then
+    destination_node = parent:named_child(found_pos + 1)
+  elseif opts.switch_parent then
+    local next_node = M.get_next_node(node:parent())
+    if next_node and next_node:named_child_count() > 0 then
+      destination_node = next_node:named_child(0)
+    elseif next_node and opts.return_next_parent then
+      destination_node = next_node
+    end
+  end
+
+  return destination_node
+end
+
+--- Get previous node of the same parent
+-- @param node The start node
+---@param opts table A table of optional arguments:
+---                  - switch_parent: (boolean, default false) Allow to return the last node from
+---                  the previous parent if the start node is the first node of its parent.
+---                  - return_prev_parent: (boolean, default false) With switch_parent = true, allow
+---                  to return the previous parent as the previous node if it has no children.
```

Comment:
I think those options are not really necessary. Those are just a relic of https://github.com/theHamsta/crazy-node-movement wouldn't know whether this is useful enough for core

Reply:
without options this should be the same as builtin `node:prev_[named_]sibling()`

Reply:
In that case, if the options are not interesting to upstream, I would drop it entirely. Same for `get_next_node`

Reply:
I think prev_sibling didn't exist at the time this was added to nvim-treesitter